source-mongodb/pull.go

```diff
@@ -96,16 +114,14 @@ func (d *driver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) err
 		}
 		var resState, _ = prevState.Resources[resourceId(res)]
 
-		// Only backfill the first time, and we know that when there is no state
-		var backfillFinishedAt *primitive.Timestamp = nil
-		if resState == nil {
-			if backfillFinishedAt, err = c.BackfillCollection(ctx, client, idx, res); err != nil {
+		if resState.Status == StatusBackfill {
+			if err = c.BackfillCollection(ctx, client, idx, res, &resState); err != nil {
 				return fmt.Errorf("backfill: %w", err)
 			}
 		}
 
 		eg.Go(func() error {
-			return c.ChangeStream(egCtx, client, idx, res, backfillFinishedAt, resState)
+			return c.ChangeStream(egCtx, client, idx, res, &resState)
```

Comment:
I think there is a race between emitting documents & checkpoints for concurrent bindings here: If there are 2 `ChangeStream` workers running concurrently, their calls to `c.Output.Documents` and `c.Output.Checkpoint` can be interleaved with each other, and also to the running `c.BackfillCollection` call. `c.Output.Documents` and `c.Output.Checkpoint` do take a lock, but it is only for their specific invocation and it is released when they return.

An example inconsistency this could lead to is:
1. Backfill collection 'A' emits some documents
2. ChangeStream collection 'B' emits some documents
3. Backfill collection 'A' emits a checkpoint; this commits the documents from (1) and (2) to flow
4. The connector crashes before ChangeStream collection 'B' emits its state checkpoint
5. When the connector starts back up, ChangeStream collection 'B' resumes from an older state checkpoint and emits duplicate documents to ones that were previously emitted and committed to flow

It could be possible to prevent this by having a kind of "emit documents and checkpoints" function that emits both under a common lock.